apps/extension/src/provider/InjectedAnoma.ts

```diff
@@ -0,0 +1,183 @@
+      postMessage: (message: any) => void;
+    } = {
+      // TODO: Remove following `any` once structure is known
+      addMessageListener: (fn: (e: any) => void) =>
+        window.addEventListener("message", fn),
+      postMessage: (message) =>
+        window.postMessage(message, window.location.origin),
+    },
+    parseMessage?: (message: any) => any
+  ) {
+    // TODO: Remove following `any` once structure is known
+    eventListener.addMessageListener(async (e: any) => {
+      const message: ProxyRequest = parseMessage
+        ? parseMessage(e.data)
+        : e.data;
+      if (!message || message.type !== "proxy-request") {
+        return;
+      }
+
+      const { method, args } = message;
+      try {
+        if (!anoma[method] || typeof anoma[method] !== "function") {
+          throw new Error(`Invalid method: ${message.method}`);
+        }
+
+        const result = await anoma[method](args);
+
+        const proxyResponse: ProxyRequestResponse = {
+          type: "proxy-request-response",
+          id: message.id,
+          result: {
+            return: result,
+          },
+        };
+
+        eventListener.postMessage(proxyResponse);
+      } catch (e: unknown) {
+        const proxyResponse: ProxyRequestResponse = {
+          type: "proxy-request-response",
+          id: message.id,
+          result: {
+            error: e,
+          },
+        };
+
+        eventListener.postMessage(proxyResponse);
+      }
+    });
+  }
+
+  protected requestMethod(
+    method: ProxyMethods,
+    args: ChainConfig & string
+    // TODO: Remove following `any` once structure is known
+  ): Promise<any> {
+    const bytes = new Uint8Array(8);
+    const id: string = Array.from(crypto.getRandomValues(bytes))
+      .map((value) => {
+        return value.toString(16);
+      })
+      .join("");
+
+    const proxyMessage: ProxyRequest = {
+      type: "proxy-request",
+      id,
+      method,
+      args,
+    };
+
+    return new Promise((resolve, reject) => {
+      // TODO: Remove following `any` once structure is known
+      const receiveResponse = (e: any) => {
+        const proxyResponse: ProxyRequestResponse = this.parseMessage
+          ? this.parseMessage(e.data)
+          : e.data;
+
+        if (!proxyResponse || proxyResponse.type !== "proxy-request-response") {
+          return;
+        }
+
+        if (proxyResponse.id !== id) {
+          return;
+        }
+
+        this.eventListener.removeMessageListener(receiveResponse);
+
+        const { result } = proxyResponse;
+
+        if (!result) {
+          reject(new Error("Result is null"));
+          return;
+        }
+
+        if (result.error) {
+          // TODO: Remove following `any` once structure is known
+          reject(new Error(result.error as any));
+          return;
+        }
+
+        resolve(result.return);
+      };
+
+      this.eventListener.addMessageListener(receiveResponse);
```

Comment:
This is what I meant, so I was wondering why we keep on adding `receiveResponse` here and then removing it inside itself once it will be called in return?

Reply:
I agree this is a little confusing (and could stand to be refactored), but essentially, we are registering a listener in `requestMethod` to validate the message we are sending back to the client. This is listening for a response from the content-scripts (which is where our API instance exists), and validating it before sending it back to the client (it must be the correct type, have the correct ID, and if any errors were encountered in the background, reject the promise and pass it back to the client).

The `proxyMessage` is constructed with the generated ID, and a handler (`receiveResponse`) is created that knows that ID (as it's in scope when this handler is defined) and can match it, and so we want to listen with that handler just before we send the response (`event.postMessage(proxyMessage)`) This is essentially adding a listener for a specific, expected message, and then immediately removing itself ([Line 117](https://github.com/anoma/namada-interface/blob/80f0e28cf1f34531b225a62bb9c9db22e2bf55a7/apps/extension/src/provider/InjectedAnoma.ts#L117). Any time we send a `ProxyMessage` to the content-scripts, we register a listener to respond to that exact message. Any message that is proxied from the client to the API gets a one-time listener for that message.

1. [L89-L93](https://github.com/anoma/namada-interface/blob/80f0e28cf1f34531b225a62bb9c9db22e2bf55a7/apps/extension/src/provider/InjectedAnoma.ts#L89-L93) - An ID is generated for the requested method to proxy
2. [L104-L133](https://github.com/anoma/namada-interface/blob/80f0e28cf1f34531b225a62bb9c9db22e2bf55a7/apps/extension/src/provider/InjectedAnoma.ts#L104-L133) - We define the listener here, which if it is able to validate that the ID of the requested method matches, and no errors have been returned, will resolve the promise and return the results to the client.